### ISSUE TITLE
hitting assert while running Socket tests on Ubuntu18

### ISSUE THREAD

**Issue by wfurt** [dotnet/runtime#42085]
I have tests running in loop on my Ubuntu18 machine (local debug build of master with release runtime)
Over run of few hundreds I hit this twice. On first occurrence I did not collect core file but I suspect same issue:

```
Process terminated. Assertion failed.
CurrentSocket and MultipleConnect both null!
SocketAsyncEventArgs#37564376.CancelConnectAsync
   at System.Net.NetEventSource.Fail(Object thisOrContextObject, Object message, String memberName) in /home/furt/github/wfurt-runtime2/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs:line 188
   at System.Net.Sockets.SocketAsyncEventArgs.CancelConnectAsync() in /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:line 581
   at System.Net.Sockets.Socket.CancelConnectAsync(SocketAsyncEventArgs e) in /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 3867
   at System.Net.Sockets.Tests.SocketAsyncEventArgsTest.CancelConnectAsync_StaticConnect_CancelsInProgressConnect() in /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketAsyncEventArgsTest.cs:line 423
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 257
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /home/furt/github/wfurt-runtime2/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:line 42
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<InvokeTestMethodAsync>b__1()
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /home/furt/github/wfurt-runtime2/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:line 42
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction)
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /home/furt/github/wfurt-runtime2/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:line 42
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code)
```

so far I failed to get managed dump with sosplugin. 
The relevant code will Assert in NetEventSource.Fail

```c#
                   // Otherwise we're doing a normal ConnectAsync - cancel it by closing the socket.
                    // _currentSocket will only be null if _multipleConnect was set, so we don't have to check.
                    if (_currentSocket == null)
                    {
                        NetEventSource.Fail(this, "CurrentSocket and MultipleConnect both null!");
                    }
                    _currentSocket.Dispose();
```

**Comment by wfurt** [comment]
This code exists for long time. But since `_currentSocket` is not seems to be protected anyhow,  what prevents the original connect finishing and setting `_currentSocket` to `null` while we are in middle of the `CancelConnectAsync`

cc: @geoffkizer

**Comment by geoffkizer** [comment]
I don't know. This does seem like a problem.

**Comment by wfurt** [comment]
I'm trying to reproduce once again and get better dump.

**Comment by wfurt** [comment]
I check and the assert was put in by @stephentoub while back.
I finally managed to get sos working and here is objdump from sos:
```
> dumpobj 0x00007f91604f31a8
Name:        System.Net.Sockets.SocketAsyncEventArgs
MethodTable: 00007f9190280058
EEClass:     00007f9190274a40
Size:        232(0xe8) bytes
File:        /home/furt/github/wfurt-runtime2/artifacts/bin/testhost/net5.0-Linux-Debug-x64/shared/Microsoft.NETCore.App/6.0.0/System.Net.Sockets.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007f918f6ee840  40002d2      200     System.EventArgs  0   static 0000000000000000 Empty
00007f918fe98cf8  40001a0        8 ...et.Sockets.Socket  0 instance 0000000000000000 _acceptSocket
00007f918fe98cf8  40001a1       10 ...et.Sockets.Socket  0 instance 0000000000000000 _connectSocket
00007f918f7a2b68  40001a2       c0 ...Private.CoreLib]]  1 instance 00007f91604f3268 _buffer
00007f918f61b170  40001a3       88         System.Int32  1 instance                0 _offset
00007f918f61b170  40001a4       8c         System.Int32  1 instance                0 _count
00007f918f617150  40001a5       b8       System.Boolean  1 instance                0 _bufferIsExplicitArray
00007f91900c62d0  40001a6       18 ...Private.CoreLib]]  0 instance 0000000000000000 _bufferList
00007f91904f9ad0  40001a7       20 ...Private.CoreLib]]  0 instance 0000000000000000 _bufferListInternal
00007f918f61b170  40001a8       90         System.Int32  1 instance                0 _bytesTransferred
00007f918f617150  40001a9       b9       System.Boolean  1 instance                0 _disconnectReuseSocket
00007f919026fd70  40001aa       94         System.Int32  1 instance                2 _completedOperation
00007f918fe9fad0  40001ab       d0 ...PacketInformation  1 instance 00007f91604f3278 _receiveMessageFromPacketInfo
00007f918fea7c50  40001ac       28  System.Net.EndPoint  0 instance 00007f91604f33f0 _remoteEndPoint
00007f918f61b170  40001ad       98         System.Int32  1 instance                0 _sendPacketsSendSize
00007f9190676d18  40001ae       30 ...dPacketsElement[]  0 instance 0000000000000000 _sendPacketsElements
00007f919026fec0  40001af       9c         System.Int32  1 instance                0 _sendPacketsFlags
00007f918fe9a920  40001b0       a0         System.Int32  1 instance            10061 _socketError
00007f918f6aa1c0  40001b1       38     System.Exception  0 instance 0000000000000000 _connectByNameError
00007f918fe9f870  40001b2       a4         System.Int32  1 instance                0 _socketFlags
00007f918f610c18  40001b3       40        System.Object  0 instance 0000000000000000 _userToken
00007f918f700bd8  40001b4       48        System.Byte[]  0 instance 0000000000000000 _acceptBuffer
00007f918f61b170  40001b5       a8         System.Int32  1 instance                0 _acceptAddressBufferCount
00007f91902879e0  40001b6       50 ...als.SocketAddress  0 instance 00007f91604f34f8 _socketAddress
00007f918f617150  40001b7       ba       System.Boolean  1 instance                1 _flowExecutionContext
00007f918f9b2aa8  40001b8       58 ....ExecutionContext  0 instance 0000000000000000 _context
00007f918fe98cf8  40001ba       60 ...et.Sockets.Socket  0 instance 0000000000000000 _currentSocket
00007f918f617150  40001bb       bb       System.Boolean  1 instance                0 _userSocket
00007f918f617150  40001bc       bc       System.Boolean  1 instance                0 _disposeCalled
00007f918f61b170  40001bd       ac         System.Int32  1 instance                0 _operating
00007f91902ac1d0  40001be       68 ...tipleConnectAsync  0 instance 0000000000000000 _multipleConnect
00007f91902805e0  40001bf       70 ...tem.Net.Sockets]]  0 instance 00007f91604f3308 Completed
00007f918f61ed78  40001c0       80        System.IntPtr  1 instance 0000000000000000 _acceptedFileDescriptor
00007f918f61b170  40001c1       b0         System.Int32  1 instance                0 _socketAddressSize
00007f918fe9f870  40001c2       b4         System.Int32  1 instance                0 _receivedFlags
00007f91904fa5c8  40001c3       78 ....Net.Primitives]]  0 instance 0000000000000000 _transferCompletionCallback
00007f918f7a8430  40001b9       d0 ...g.ContextCallback  0   static 00007f91602dd320 s_executionCallback
```


so while both _currentSocket and _multipleConnect are both null, `_operating` is 0 e.g. `Free`. Since this whole block is guarded by 
```c#
if (_operating == InProgress ...
```
it is clear IMHO that the values changed between the initial check and body of the if. Or the `_operating` did get stale because of some optimization. Feels similar to #42013 - something perhaps wrong for a while but exposed because of some recent changes. 

It looks like a product bug to me - not a bad test.

**Comment by wfurt** [comment]
It seems like this can also throw NRE if _currentSocket changes between check and calling Dispose on it.
```
   System.Net.Sockets.Tests.DualModeConnectAsync.DualModeConnectAsync_DnsEndPointToHost_Helper [SKIP]
      Condition(s) not met: "LocalhostIsBothIPv4AndIPv6"
    System.Net.Sockets.Tests.CreateSocket.Ctor_Raw_Supported_Success [SKIP]
      Condition(s) not met: "SupportsRawSockets"
      System.NullReferenceException : Object reference not set to an instance of an object.
      Stack Trace:
        /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs(583,0): at System.Net.Sockets.SocketAsyncEventArgs.CancelConnectAsync()
        /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(3867,0): at System.Net.Sockets.Socket.CancelConnectAsync(SocketAsyncEventArgs e)
        /home/furt/github/wfurt-runtime2/src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketAsyncEventArgsTest.cs(423,0): at System.Net.Sockets.Tests.SocketAsyncEventArgsTest.CancelConnectAsync_StaticConnect_CancelsInProgressConnect()
  Finished:    System.Net.Sockets.Tests
Canceling due to test failure...
=== TEST EXECUTION SUMMARY ===
   System.Net.Sockets.Tests  Total: 1329, Errors: 0, Failed: 1, Skipped: 5, Time: 15.743s
Failed in loop 314
```

**Comment by wfurt** [comment]
fixed by #42935
